core/src/apps/cardano/sign_tx.py

```diff
@@ -125,117 +109,82 @@ async def sign_tx(ctx, msg, keychain: seed.Keychain):
 class Transaction:
     def __init__(
         self,
-        inputs: list,
-        outputs: list,
-        keychain,
+        inputs: List[CardanoTxInputType],
+        outputs: List[CardanoTxOutputType],
+        keychain: seed.Keychain,
         protocol_magic: int,
-        input_coins_sum: int,
-    ):
+        fee: int,
+        ttl: int,
+    ) -> None:
         self.inputs = inputs
         self.outputs = outputs
         self.keychain = keychain
-        # attributes have to be always empty in current Cardano
-        self.attributes = {}
+        self.fee = fee
+        self.ttl = ttl
 
         self.network_name = KNOWN_PROTOCOL_MAGICS.get(protocol_magic, "Unknown")
         self.protocol_magic = protocol_magic
-        self.input_coins_sum = input_coins_sum
 
-    def _process_outputs(self):
-        change_addresses = []
-        change_derivation_paths = []
-        output_addresses = []
-        outgoing_coins = []
-        change_coins = []
+    def _build_inputs(self) -> list:
+        inputs_for_cbor = []
+        for input in self.inputs:
+            inputs_for_cbor.append([input.prev_hash, input.prev_index])
+
+        return inputs_for_cbor
+
+    def _build_outputs(self) -> list:
+        outputs_for_cbor = []
 
         for output in self.outputs:
+            amount = output.amount
             if output.address_n:
-                address, _ = derive_address_and_node(self.keychain, output.address_n)
-                change_addresses.append(address)
-                change_derivation_paths.append(output.address_n)
-                change_coins.append(output.amount)
+                address, _ = derive_address_and_node(
+                    self.keychain, output.address_n, self.protocol_magic
+                )
+                # set output address so it can later be shown to the user if needed
+                output.address = address
             else:
-                if output.address is None:
+                address = output.address
+                if address is None:
                     raise wire.ProcessError(
                         "Each output must have address or address_n field!"
                     )
-                if not is_safe_output_address(output.address):
+                if not is_safe_output_address(address):
                     raise wire.ProcessError("Invalid output address!")
 
-                outgoing_coins.append(output.amount)
-                output_addresses.append(output.address)
+            outputs_for_cbor.append([base58.decode(address), amount])
 
-        self.change_addresses = change_addresses
-        self.output_addresses = output_addresses
-        self.outgoing_coins = outgoing_coins
-        self.change_coins = change_coins
-        self.change_derivation_paths = change_derivation_paths
+        return outputs_for_cbor
 
-    def _build_witnesses(self, tx_aux_hash: str):
+    def _build_witnesses(self, tx_aux_hash: bytes) -> List[List[bytes]]:
```

Comment:
Now I'm not sure, but have we removed Byron witness computation? That does not sound right, we still need it for legacy utxos, don't we?

Reply:
We haven't. We simply don't know for sure what it will be, so I have decided to leave it out until we do.